### PR TITLE
FDM solver for the de Rham complex

### PR DESCRIPTION
**Pull request by pbrubeck**
This PR enables `FDMPC` to discretize a sparse auxiliary operator for de Rham-conforming discretizations on quads and hexes using elements with `variant="fdm"`. 

The previous behaviour of `FDMPC` targeted an IPDG discretization of a Poisson-like problem (grad u, alpha grad v) + (u, beta v) on non H1-conforming elements. This is still accessible through `PoissonFDMPC` and using elements with `variant="fdm_ipdg"`.

### CODE REVIEW

File: firedrake/preconditioners/pmg.py [hunk 1]
```diff
@@ -499,6 +504,89 @@ def coarsen_bc_value(self, bc, cV):
...
+    cppargs = ["-I%s/include" % d for d in get_petsc_dir()]
+    ldargs = (["-L%s/lib" % d for d in get_petsc_dir()]
+              + ["-Wl,-rpath,%s/lib" % d for d in get_petsc_dir()]
+              + ["-lpetsc", "-lm"])
+    return load(code, "c", name, argtypes=argtypes,
+                cppargs=cppargs, ldargs=ldargs,
+                comm=comm)
+
+
+def reference_moments(*args, **kwargs):
```
Comment (pefarrell) [inline firedrake/preconditioners/pmg.py:523]: Can you add a comment explaining what this is doing?

File: firedrake/preconditioners/pmg.py [hunk 2]
```diff
@@ -515,25 +603,27 @@ def prolongation_transfer_kernel_action(Vf, expr):
...
 
 
-@lru_cache(maxsize=10)
 def expand_element(ele):
     """
     Expand a FiniteElement as an EnrichedElement of TensorProductElements, discarding modifiers.
     """
-    if ele.cell() == ufl.quadrilateral:
+
+    if ele.cell().cellname().startswith("quadrilateral"):
```
Comment (pefarrell) [inline firedrake/preconditioners/pmg.py:611]: Why the `startswith`? Can you add a comment?

File: firedrake/preconditioners/pmg.py [hunk 3]
```diff
@@ -675,42 +899,206 @@ def get_line_interpolator(felem, celem):
...
-        raise ValueError("Fine and coarse elements do not have the same number of factors")
-    if len(felems) > 3:
-        raise ValueError("More than three factors are not supported")
 
-    # Declare array shapes to be used as literals inside the kernels
-    fshape = [e.space_dimension() for e in felems]
-    cshape = [e.space_dimension() for e in celems]
-    shapes = [(nscal,) + tuple(fshape), (nscal,) + tuple(cshape)]
+@PETSc.Log.EventDecorator("MakeKronCode")
+def make_kron_code(Vf, Vc, t_in, t_out, mat_name, scratch):
```
Comment (pefarrell) [inline firedrake/preconditioners/pmg.py:969]: Can you please explain what all of this is about? It's pretty opaque.

File: firedrake/preconditioners/pmg.py [hunk 4]
```diff
@@ -850,80 +1232,134 @@ def make_permutation_code(V, vshape, pshape, t_in, t_out, array_name):
...
+
     def __init__(self, Vf, Vc, Vf_bcs, Vc_bcs):
         self.Vf_bcs = Vf_bcs
         self.Vc_bcs = Vc_bcs
         if isinstance(Vf, firedrake.Function):
             self.uf = Vf
             Vf = Vf.function_space()
         else:
-            self.uf = firedrake.Function(Vf)
+            self.uf = self._cache_work.get(Vf, firedrake.Function(Vf))
```
Comment (pefarrell) [inline firedrake/preconditioners/pmg.py:1274]: This will always malloc a Function, even if there is a cache hit.

File: tests/regression/test_fdm.py [hunk 5]
```diff
@@ -95,9 +94,50 @@ def test_p_independence(mesh, expected, variant):
         problem = LinearVariationalProblem(a, L, uh, bcs=bcs)
         solver = LinearVariationalSolver(problem, solver_parameters=fdmstar)
         solver.solve()
-        nits.append(solver.snes.ksp.getIterationNumber())
+        assert solver.snes.ksp.getIterationNumber() <= nits
```
Comment (pefarrell) [inline tests/regression/test_fdm.py:97]: How come the number of expected iterations didn't change as you solved new PDEs? Is it a coincidence that the iteration counts are the same?

File: tests/regression/test_fdm.py [hunk 6]
```diff
@@ -255,3 +295,48 @@ def test_direct_solver(fs):
 
     assert solver.snes.ksp.getIterationNumber() == 1
     assert norm(u_exact-uh, "H1") < 1.0E-8
+
+
+@pytest.mark.skipcomplex
+def test_static_condensation(mesh):
```
Comment (pefarrell) [inline tests/regression/test_fdm.py:301]: This doesn't seem related to the main thrust of this PR?

**Review by pefarrell — CHANGES_REQUESTED** [review]

File: firedrake/preconditioners/pmg.py [hunk 7]
```diff
@@ -121,8 +129,8 @@ def initialize(self, pc):
         pdm.setCreateInterpolation(self.create_interpolation)
         # We need this for p-FAS
         pdm.setCreateInjection(self.create_injection)
-        pdm.setSNESJacobian(_SNESContext.form_jacobian)
         pdm.setSNESFunction(_SNESContext.form_function)
+        pdm.setSNESJacobian(_SNESContext.form_jacobian)
```
[the remark this answers is not in the document]
Comment (pbrubeck) [reply firedrake/preconditioners/pmg.py:133]: no reason at al

File: tests/regression/test_fdm.py (hunk 6 above)
Comment (pbrubeck) [reply tests/regression/test_fdm.py:301]: FDMPC automatically statically condenses the problem if given a facet element, this is just testing SLATE's matfree interior solvers and using FDMPC on the Schur complement.

Comment (pbrubeck) [reply tests/regression/test_fdm.py:301]: There is another similar test that I might need to include, because FDMPC is supposed to handle MixedFunctionSpaces that arise from an interior-facet decomposition.

File: firedrake/preconditioners/fdm.py [hunk 8]
```diff
@@ -136,73 +170,919 @@ def interp_nullspace(I, nsp):
...
+            self.lgmaps[Vsub] = lgmap
+
+            own = Vsub.dof_dset.layout_vec.getLocalSize()
+            bdofs = numpy.nonzero(lgmap.indices[:own] < 0)[0].astype(PETSc.IntType)
+            bc_rows[Vsub] = Vsub.dof_dset.lgmap.apply(bdofs, result=bdofs)
+
+        # get coefficients on a given cell
+        coefficients, assembly_callables = self.assemble_coef(J, form_compiler_parameters)
+        coeffs = [coefficients.get(k) for k in ("beta", "alpha")]
+        cmaps = [glonum_fun(ck.cell_node_map())[0] for ck in coeffs]
```
Comment (pefarrell) [inline firedrake/preconditioners/fdm.py:259]: Is there a more intuitive variable name than `glonum_fun`?

File: firedrake/preconditioners/fdm.py [hunk 9]
```diff
@@ -136,73 +170,919 @@ def interp_nullspace(I, nsp):
...
+
+        @PETSc.Log.EventDecorator("FDMGetCoeffs")
+        def get_coeffs(e, result=None):
+            vals = []
```
Comment (pefarrell) [inline firedrake/preconditioners/fdm.py:263]: Comment "Get vector for betas and alphas on a cell"

File: firedrake/preconditioners/fdm.py [hunk 10]
```diff
@@ -136,73 +170,919 @@ def interp_nullspace(I, nsp):
...
+        self.nel = nel
+        self.work_mats = dict()
+
+        Pmats = dict()
+        addv = PETSc.InsertMode.ADD_VALUES
+        symmetric = pmat_type.endswith("sbaij")
+
+        # Store only off-diagonal blocks with more columns than rows to save memory
+        Vsort = sorted(V, key=lambda Vsub: Vsub.dim())
+        for Vrow, Vcol in product(Vsort, Vsort):
```
Comment (pefarrell) [inline firedrake/preconditioners/fdm.py:280]: Comment "Loop over all pairs of subspaces"

File: firedrake/preconditioners/fdm.py [hunk 11]
```diff
@@ -136,73 +170,919 @@ def interp_nullspace(I, nsp):
...
+        if hasattr(self, "work_mats"):
+            objs.extend(list(self.work_mats.values()))
+        if hasattr(self, "ises"):
+            objs.extend(self.ises)
+        for obj in objs:
+            if hasattr(obj, "destroy"):
+                obj.destroy()
+
+    @PETSc.Log.EventDecorator("FDMSetValues")
+    def set_values(self, A, Vrow, Vcol, addv, triu=False):
```
Comment (pefarrell) [inline firedrake/preconditioners/fdm.py:381]: Comment

File: firedrake/preconditioners/fdm.py [hunk 12]
```diff
@@ -136,73 +170,919 @@ def interp_nullspace(I, nsp):
...
+            return R.transposeMatMult(RtAP.buff, result=result)
+        RtAP.buff = None
+
+        set_values_csr = self.load_set_values(triu=triu)
+        get_rindices = self.cell_to_global[Vrow]
+        if Vrow == Vcol:
+            get_cindices = lambda e, result=None: result
+            update_A = lambda Ae, rindices, cindices: set_values_csr(A, Ae, rindices, rindices, addv)
+            rtensor = self.reference_tensor_on_diag.get(Vrow) or self.assemble_reference_tensor(Vrow)
+            assemble_element_mat = lambda De, result=None: De.PtAP(rtensor, result=result)
```
Comment (pefarrell) [inline firedrake/preconditioners/fdm.py:394]: Refer to formula such-and-such in the manuscript

File: firedrake/preconditioners/fdm.py [hunk 13]
```diff
@@ -136,73 +170,919 @@ def interp_nullspace(I, nsp):
...
+                    De.assemble()
+                    return De
+            else:
+                @PETSc.Log.EventDecorator("FDMUpdateDiag")
+                def update_De(data):
+                    work_vec.setArray(data)
+                    De.setDiagonal(work_vec, addv=insert)
+                    return De
+
+            for e in range(self.nel):
```
Comment (pefarrell) [inline firedrake/preconditioners/fdm.py:431]: Comment "Core assembly loop"

File: firedrake/preconditioners/fdm.py [hunk 14]
```diff
@@ -136,73 +170,919 @@ def interp_nullspace(I, nsp):
...
+                rindices = get_rindices(e, result=rindices)
+                cindices = get_cindices(e, result=cindices)
+                data = self.get_coeffs(e, result=data)
+                Ae = assemble_element_mat(update_De(data), result=Ae)
+                update_A(condense_element_mat(Ae), rindices, cindices)
+
+            work_vec.destroy()
+
+        elif self.nel:
```
Comment (pefarrell) [inline firedrake/preconditioners/fdm.py:440]: Comment "This code is for preallocation of the sparsity pattern"

File: firedrake/preconditioners/fdm.py [hunk 15]
```diff
@@ -136,73 +170,919 @@ def interp_nullspace(I, nsp):
...
+            self.work_mats[common_key] = None
+            self.work_mats[Vrow, Vcol] = None
+        if RtAP.buff:
+            RtAP.buff.destroy()
+
+    @PETSc.Log.EventDecorator("FDMCoefficients")
+    def assemble_coef(self, J, form_compiler_parameters):
         """
-        Assemble the sparse preconditioner with cell-wise constant coefficients.
+        Obtain coefficients as the diagonal of a weighted mass matrix in V^k x V^{k+1}
```
Comment (pefarrell) [inline firedrake/preconditioners/fdm.py:477]: Refer to discussion in manuscript.

Explain basic strategy for how these will be extracted.

File: firedrake/preconditioners/fdm.py [hunk 16]
```diff
@@ -136,73 +170,919 @@ def interp_nullspace(I, nsp):
...
+                assembly_callables = []
+                for iset, name in zip(Z.dof_dset.field_ises, ("beta", "alpha")):
+                    sub = M.petscmat.createSubMatrix(iset, iset)
+                    ctx = sub.getPythonContext()
+                    coefficients[name] = ctx._block_diagonal
+                    assembly_callables.append(ctx._assemble_block_diagonal)
+            return self._coefficient_cache.setdefault(key, (coefficients, assembly_callables))
+
+    @PETSc.Log.EventDecorator("FDMRefTensor")
+    def assemble_reference_tensor(self, V):
```
Comment (pefarrell) [inline firedrake/preconditioners/fdm.py:569]: Refer to section of manuscript.

Explain basic idea.

File: firedrake/preconditioners/fdm.py [hunk 17]
```diff
@@ -136,73 +170,919 @@ def interp_nullspace(I, nsp):
...
+            if is_facet:
+                cache[full_key] = result
+                noperm = PETSc.IS().createGeneral(numpy.arange(result.getSize()[0], dtype=PETSc.IntType), comm=result.comm)
+                result = result.createSubMatrix(noperm, self.ises[1])
+                noperm.destroy()
+
+            return cache.setdefault(key, result)
+
+
+def factor_interior_mat(A00):
```
Comment (pefarrell) [inline firedrake/preconditioners/fdm.py:642]: Comment "Used in static condensation. Take in A00 on a cell, return its Cholesky factorisation."

File: firedrake/preconditioners/fdm.py [hunk 18]
```diff
@@ -136,73 +170,919 @@ def interp_nullspace(I, nsp):
...
+    # Assume that interior DOF list i0 is ordered such that A00 is block diagonal
+    # with blocks of increasing dimension
+    indptr, indices, data = A00.getValuesCSR()
+    degree = numpy.diff(indptr)
+
+    # TODO handle non-symmetric case with LU, requires scipy
```
Comment (pefarrell) [inline firedrake/preconditioners/fdm.py:648]: Raise an error if not symmetric

File: firedrake/preconditioners/fdm.py [hunk 19]
```diff
@@ -136,73 +170,919 @@ def interp_nullspace(I, nsp):
...
+        data[zlice] = invchol(data[zlice].reshape((-1, k, k))).reshape((-1,))
+        flops = ((k+1)**3 + 5*(k+1)-12)//3 + k**3
+        PETSc.Log.logFlops(flops*nblocks)
+
+    A00.setValuesCSR(indptr, indices, data)
+    A00.assemble()
+
+
+@PETSc.Log.EventDecorator("FDMCondense")
+def condense_element_mat(A, i0, i1, submats):
```
Comment (pefarrell) [inline firedrake/preconditioners/fdm.py:667]: Comment "Return the Schur complement associated to indices in i1, condensing i0 out"

File: firedrake/preconditioners/fdm.py [hunk 20]
```diff
@@ -136,73 +170,919 @@ def interp_nullspace(I, nsp):
...
+            return obj.ctypes.data
+        return obj
+
+    @PETSc.Log.EventDecorator(name)
+    def wrapper(*args):
+        return funptr(*list(map(get_pointer, args)))
+    return wrapper
+
+
+def load_assemble_csr(comm, triu=False):
```
Comment (pefarrell) [inline firedrake/preconditioners/fdm.py:718]: Comment "Insert one sparse matrix into another sparse matrix. Done in C for efficiency, since it loops over rows."

File: firedrake/preconditioners/fdm.py [hunk 21]
```diff
@@ -136,73 +170,919 @@ def interp_nullspace(I, nsp):
...
+    PetscFunctionReturn(0);
+}}
+"""
+    argtypes = [ctypes.c_voidp, ctypes.c_voidp,
+                ctypes.c_voidp, ctypes.c_voidp, ctypes.c_int]
+    return load_c_code(code, name, comm=comm, argtypes=argtypes,
+                       restype=ctypes.c_int)
+
+
+def petsc_sparse(A_numpy, rtol=1E-10):
```
Comment (pefarrell) [inline firedrake/preconditioners/fdm.py:770]: Comment "Convert dense numpy matrix into a sparse PETSc matrix"

File: firedrake/preconditioners/fdm.py [hunk 22]
```diff
@@ -136,73 +170,919 @@ def interp_nullspace(I, nsp):
...
+    nnz = numpy.count_nonzero(abs(A_numpy) > atol, axis=1).astype(PETSc.IntType)
+    A = PETSc.Mat().createAIJ(A_numpy.shape, nnz=(nnz, 0), comm=PETSc.COMM_SELF)
+    for row, Arow in enumerate(A_numpy):
+        cols = numpy.argwhere(abs(Arow) > atol).astype(PETSc.IntType).flat
+        A.setValues(row, cols, Arow[cols], PETSc.InsertMode.INSERT)
+    A.assemble()
+    return A
+
+
+def block_mat(A_blocks):
```
Comment (pefarrell) [inline firedrake/preconditioners/fdm.py:782]: Comment "A_blocks is an array of arrays defining a block matrix of PETSc Mats. Returns the corresponding overall PETSc Mat. We make an AIJ instead of a nest because ..."

File: firedrake/preconditioners/fdm.py [hunk 23]
```diff
@@ -136,73 +170,919 @@ def interp_nullspace(I, nsp):
...
+    cols = [numpy.arange(*csizes[j:j+2], dtype=PETSc.IntType) for j in range(len(A_blocks[0]))]
+    for Ai, irows in zip(A_blocks, rows):
+        for Aij, jcols in zip(Ai, cols):
+            insert_block(A, Aij, irows, jcols, imode)
+
+    A.assemble()
+    return A
+
+
+def is_restricted(finat_element):
```
Comment (pefarrell) [inline firedrake/preconditioners/fdm.py:805]: Comment "Determine if an element is a restriction onto interior or facets"

File: firedrake/preconditioners/fdm.py [hunk 24]
```diff
@@ -136,73 +170,919 @@ def interp_nullspace(I, nsp):
...
+    A11.destroy()
+    A10.destroy()
+    result = block_mat(A_blocks)
+    for A_row in A_blocks:
+        for A in A_row:
+            A.destroy()
+    return result
+
+
+def diff_prolongator(Vf, Vc, fbcs=[], cbcs=[]):
```
Comment (pefarrell) [inline firedrake/preconditioners/fdm.py:934]: Comment "Magic. Tabulate grad: H(grad) -> H(curl), or curl: H(curl) -> H(div), etc as an explicit sparse matrix. Works for any basis. These are the same matrices one needs for HypreAMS and friends."

File: firedrake/preconditioners/pmg.py [hunk 25]
```diff
@@ -195,6 +206,7 @@ def _coarsen_form(a):
         cJp = _coarsen_form(fctx.Jp)
         fcp = self.coarsen_quadrature(fproblem.form_compiler_parameters, fdeg, cdeg)
         cbcs = self.coarsen_bcs(fproblem.bcs, cV)
+        cF = self.coarsen_residual(cF, cJ, cu)
```
[the remark this answers is not in the document]
Comment (pefarrell) [reply firedrake/preconditioners/pmg.py:209]: Maybe this function should be renamed?

[the remark this answers is not in the document]
Comment (pefarrell) [reply firedrake/preconditioners/pmg.py:209]: Maybe just do it inline?

File: firedrake/preconditioners/pmg.py [hunk 26]
```diff
@@ -507,100 +523,177 @@ def prolongation_transfer_kernel_action(Vf, expr):
...
-        else:
-            return ufl.EnrichedElement(*[ufl.TensorProductElement(*k) for k in terms])
+        terms = list(map(finat.TensorProductElement, terms))
+        return finat.EnrichedElement(terms)
     else:
         return ele
 
 
-def get_line_elements(V):
+def evaluate_dual(dual, element, key=None):
```
[the remark this answers is not in the document]
Comment (pefarrell) [reply firedrake/preconditioners/pmg.py:562]: Question for @wence- @dham @rckirby : is it possible to do this neatly with FIAT?

File: firedrake/preconditioners/pmg.py [hunk 27]
```diff
@@ -507,100 +523,177 @@ def prolongation_transfer_kernel_action(Vf, expr):
...
+
+def compare_dual_basis(l1, l2):
+    if len(l1) != len(l2):
+        return False
+    return all(compare_dual(b1, b2) for b1, b2 in zip(l1, l2))
+
+
+@lru_cache(maxsize=10)
+@PETSc.Log.EventDecorator("GetLineElements")
+def get_permutation_to_line_elements(finat_element):
```
Comment (pefarrell) [inline firedrake/preconditioners/pmg.py:611]: Add a comment explaining what this  does and why it lives in pmg.py.

File: firedrake/preconditioners/pmg.py [hunk 28]
```diff
@@ -675,42 +768,226 @@ def get_line_interpolator(felem, celem):
...
-            kronmxv(0, {fargs}, {cargs}, {nscal}, {Jargs}, &{t_in}, &{t_out});
+    kronmxv_inplace(tflag, mx, my, mz, nx, ny, nz, nel, A1, A2, A3, &ptr[0], &ptr[1]);
+
+    if(ptr[1] != y)
+        for(PetscBLASInt j=0; j<mx*my*mz*nel; j++)
+            y[j] = ptr[1][j];
+    return;
+}
+
+static inline void permute_axis(PetscBLASInt axis,
```
Comment (pefarrell) [inline firedrake/preconditioners/pmg.py:792]: Explain what this does and why you need it.

File: firedrake/preconditioners/pmg.py [hunk 29]
```diff
@@ -675,42 +768,226 @@ def get_line_interpolator(felem, celem):
...
+    }
+    for(PetscBLASInt i3=0; i3<n3; i3++)
+        for(PetscBLASInt i2=0; i2<n2; i2++)
+            for(PetscBLASInt i1=0; i1<n1; i1++)
+                for(PetscBLASInt i0=0; i0<n0; i0++)
+                    y[s0*i0 + s1*i1 + s2*i2 + s3*i3] = x[p++];
+    return;
+}
+
+static inline void ipermute_axis(PetscBLASInt axis,
```
Comment (pefarrell) [inline firedrake/preconditioners/pmg.py:813]: Same again.

**Review by pefarrell — CHANGES_REQUESTED** [review]

File: firedrake/preconditioners/pmg.py [hunk 30]
```diff
@@ -132,20 +141,22 @@ def initialize(self, pc):
...
-        self.ppc.setUp()
+        ppc.incrementTabLevel(1, parent=obj)
+        ppc.setFromOptions()
+        ppc.setUp()
+        self.ppc = ppc
 
-    def update(self, pc):
-        pass
+    def update(self, obj):
+        self.ppc.setUp()
```
[the remark this answers is not in the document]
Comment (pbrubeck) [reply firedrake/preconditioners/pmg.py:150]: The only multigrid data structures that need to be updated are the smoothers and the coarse grid solver, and these always seemed to be updated properly. It is a bit mysterious how these MG components get updated without this call.

File: firedrake/preconditioners/pmg.py [hunk 31]
```diff
@@ -507,100 +523,169 @@ def prolongation_transfer_kernel_action(Vf, expr):
...
+        return finat.EnrichedElement(terms)
     else:
         return ele
 
 
-def get_line_elements(V):
+def evaluate_dual(dual, element, key=None):
+    # Evaluate the action of a set of dual functionals on the basis functions of an element.
+    keys = set(tuple(phi.get_point_dict().keys()) for phi in dual)
+    pts = list(set(sum(keys, ())))
```
Comment (wence-) [inline firedrake/preconditioners/pmg.py:563]: These are not deterministically ordered, does it matter?

File: firedrake/preconditioners/fdm.py (hunk 18 above)
Comment (pbrubeck) [reply firedrake/preconditioners/fdm.py:685]: This might have some impact in performance

File: firedrake/preconditioners/pmg.py (hunk 31 above)
Comment (pbrubeck) [reply firedrake/preconditioners/pmg.py:563]: I believe that since we are integrating by summing over quadrature points the order of summation might have an effect on the result, causing different kernel code to be compiled on different processes. I guess `numpy.unique` will DTRT

File: firedrake/preconditioners/fdm.py [hunk 32]
```diff
@@ -136,73 +178,971 @@ def interp_nullspace(I, nsp):
...
+    submats[4] = A10.matTransposeMult(A00, result=submats[4])
+    submats[5] = A00.matMult(A01, result=submats[5])
+    submats[6] = submats[4].matMult(submats[5], result=submats[6])
+    submats[6].aypx(0.0, A, structure=structure)
+    return submats[6]
+
+
+@PETSc.Log.EventDecorator("LoadCode")
+def load_c_code(code, name, **kwargs):
+    from pyop2.compilation import load
```
Comment (pbrubeck) [inline firedrake/preconditioners/fdm.py:750]: move imports to the top

File: firedrake/preconditioners/fdm.py [hunk 33]
```diff
@@ -136,73 +178,971 @@ def interp_nullspace(I, nsp):
...
+def load_assemble_csr(comm, triu=False):
+    # Insert one sparse matrix into another sparse matrix.
+    # Done in C for efficiency, since it loops over rows.
+    if triu:
+        name = "setSubMatCSR_SBAIJ"
+        select_cols = "icol < irow ? -1: icol"
+    else:
+        name = "setSubMatCSR_AIJ"
+        select_cols = "icol"
+    code = f"""
```
Comment (pbrubeck) [inline firedrake/preconditioners/fdm.py:782]: write a small test

File: firedrake/preconditioners/fdm.py [hunk 34]
```diff
@@ -1,128 +1,179 @@
...
-        if len(bcs) > 0:
-            self.bc_nodes = numpy.unique(numpy.concatenate([bcdofs(bc, ghost=False) for bc in bcs]))
-        else:
-            self.bc_nodes = numpy.empty(0, dtype=PETSc.IntType)
+            # Construct interpolation from variant to original spaces
+            self.fdm_interp = prolongation_matrix_matfree(V_fdm, V, bcs_fdm, [])
+            self.work_vec_x = Amat.createVecLeft()
+            self.work_vec_y = Amat.createVecRight()
+            if use_amat:
+                from firedrake.assemble import allocate_matrix, TwoFormAssembler
```
Comment (connorjward) [inline firedrake/preconditioners/fdm.py:161]: Could these imports be at the top of the file? They should only be here to avoid cyclic import issues.

File: firedrake/preconditioners/fdm.py [hunk 35]
```diff
@@ -136,73 +187,960 @@ def interp_nullspace(I, nsp):
...
+            self.get_static_condensation[V] = lambda Ae: condense_element_pattern(Ae, self.ises[0], i1, self.submats)
+
+        @PETSc.Log.EventDecorator("FDMGetIndices")
+        def cell_to_global(lgmap, cell_to_local, cell_index, result=None):
+            # Be careful not to create new arrays
+            result = cell_to_local(cell_index, result=result)
+            return lgmap.apply(result, result=result)
+
+        # Create data structures needed for assembly
+        self.cell_to_global = dict()
```
Comment (connorjward) [inline firedrake/preconditioners/fdm.py:259]: Why not `self.cell_to_global = {}` (and others). I would say that it's more Pythonic.

File: firedrake/preconditioners/fdm.py [hunk 36]
```diff
@@ -136,73 +187,960 @@ def interp_nullspace(I, nsp):
...
+        coefficients, assembly_callables = self.assemble_coef(J, form_compiler_parameters)
+        coeffs = [coefficients.get(k) for k in ("beta", "alpha")]
+        cmaps = [extrude_node_map(ck.cell_node_map())[0] for ck in coeffs]
+
+        @PETSc.Log.EventDecorator("FDMGetCoeffs")
+        def get_coeffs(e, result=None):
+            # Get vector for betas and alphas on a cell
+            vals = []
+            for k, (coeff, cmap) in enumerate(zip(coeffs, cmaps)):
+                get_coeffs.indices[k] = cmap(e, result=get_coeffs.indices[k])
```
Comment (connorjward) [inline firedrake/preconditioners/fdm.py:282]: This looks like you are attaching attributes to a function pointer from within the function itself. Can you explain what is going on? I think there may be a clearer way to achieve what you want.

File: firedrake/preconditioners/fdm.py [hunk 37]
```diff
@@ -136,73 +187,960 @@ def interp_nullspace(I, nsp):
...
+        :arg A: the :class:`PETSc.Mat` to assemble
+        :arg Vrow: the :class:`.FunctionSpace` test space
+        :arg Vcol: the :class:`.FunctionSpace` trial space
+        :arg addv: a `PETSc.Mat.InsertMode`
+        :arg triu: are we assembling only the upper triangular part?
+        """
+        def RtAP(R, A, P, result=None):
+            RtAP.buff = A.matMult(P, result=RtAP.buff)
+            return R.transposeMatMult(RtAP.buff, result=result)
+        RtAP.buff = None
```
Comment (connorjward) [inline firedrake/preconditioners/fdm.py:415]: Same comment about function pointer attributes as above.

File: firedrake/preconditioners/fdm.py [hunk 38]
```diff
@@ -136,73 +187,960 @@ def interp_nullspace(I, nsp):
...
+        return obj
+
+    @PETSc.Log.EventDecorator(name)
+    def wrapper(*args):
+        return funptr(*list(map(get_pointer, args)))
+    return wrapper
+
+
+def load_assemble_csr(comm, triu=False):
+    # Insert one sparse matrix into another sparse matrix.
```
[the remark this answers is not in the document]
Comment (connorjward) [reply firedrake/preconditioners/fdm.py:776]: Also the function name does not seem to correspond to what it does. What about `merge_matrices` or `insert_submat` or something?

**Comment by pbrubeck** [comment]
@JDBetteridge here's an ascii_flamegraph for one of the slowest tests (which is not new to this PR) `tests/regression/test_fdm.py::test_ipdg_direct_solver[rt-Box]`

image


The test takes over 2 minutes, and 84% of the time is spent in kernel compilation.
The memory profiler indicates a peak memory size of 259 MB only for this test.

[test_ipdg_direct_solver[rt-Box].txt](https://github.com/firedrakeproject/firedrake/files/11050657/test_ipdg_direct_solver.rt-Box.txt)

File: firedrake/preconditioners/pmg.py [hunk 39]
```diff
@@ -855,75 +1139,129 @@ def get_permuted_map(V):
...
+        if viewer is None:
+            return
+        typ = viewer.getType()
+        if typ != PETSc.Viewer.Type.ASCII:
+            return
+        viewer.printfASCII("Firedrake matrix-free prolongator %s\n" %
+                           type(self).__name__)
+
+    def getInfo(self, mat, info=None):
+        from mpi4py import MPI
```
Comment (connorjward) [inline firedrake/preconditioners/pmg.py:1240]: Does this import need to be here?

**Review by connorjward — DISMISSED** [review]

**Comment by pbrubeck** [comment]
> > @JDBetteridge @connorjward here's an ascii_flamegraph for one of the slowest tests (which is not new to this PR) `tests/regression/test_fdm.py::test_ipdg_direct_solver[rt-Box]`
> > image
> > The test takes over 2 minutes, and 84% of the time is spent in kernel compilation. The memory profiler indicates a peak memory size of 259 MB only for this test.
> > [test_ipdg_direct_solver[rt-Box].txt](https://github.com/firedrakeproject/firedrake/files/11050657/test_ipdg_direct_solver.rt-Box.txt)
> 
> Okay, I'm happy that the test isn't using a ridiculous amount of memory as I feared. However, the time spent compiling (especially if you're seeing it with hot caches!) looks like an issue. We can discuss this in person as planned

This is not using hot caches.

Right now I am trying to reduce the number of kernels to compile by specifying a RHS of the form `action(a, u)`  where `u = Function(V).project(u_exact)`. Apparently firedrake's caching powers do not help for this case.

**Comment by connorjward** [comment]
> > > @JDBetteridge @connorjward here's an ascii_flamegraph for one of the slowest tests (which is not new to this PR) `tests/regression/test_fdm.py::test_ipdg_direct_solver[rt-Box]`
> > > image
> > > The test takes over 2 minutes, and 84% of the time is spent in kernel compilation. The memory profiler indicates a peak memory size of 259 MB only for this test.
> > > [test_ipdg_direct_solver[rt-Box].txt](https://github.com/firedrakeproject/firedrake/files/11050657/test_ipdg_direct_solver.rt-Box.txt)
> > 
> > 
> > Okay, I'm happy that the test isn't using a ridiculous amount of memory as I feared. However, the time spent compiling (especially if you're seeing it with hot caches!) looks like an issue. We can discuss this in person as planned
> 
> This is not using hot caches.
> 
> Right now I am trying to reduce the number of kernels to compile by specifying a RHS of the form `action(a, u)` where `u = Function(V).project(u_exact)`. Apparently firedrake's caching powers do not help for this case.

With a hot cache this drops from 2 minutes to 5 seconds. If I'm understanding you right, you are saying that we are compiling more `.so` files than we need to because we are not recognising that multiple kernels are doing the same thing. If that's the case then could you produce an MFE? Then we can figure out which part of the cache key is overly restrictive.

**Comment by pbrubeck** [comment]
> With a hot cache this drops from 2 minutes to 5 seconds. If I'm understanding you right, you are saying that we are compiling more `.so` files than we need to because we are not recognising that multiple kernels are doing the same thing. If that's the case then could you produce an MFE? Then we can figure out which part of the cache key is overly restrictive.

[Here](https://github.com/firedrakeproject/firedrake/issues/2841) it is, thanks!

File: firedrake/preconditioners/fdm.py (hunk 34 above)
Comment (pbrubeck) [reply firedrake/preconditioners/fdm.py:164]: This is required to avoid a cyclic import.

Comment (pbrubeck) [reply firedrake/preconditioners/fdm.py:164]: This is required to avoid a cyclic import.

File: firedrake/preconditioners/fdm.py (hunk 37 above)
Comment (pbrubeck) [reply firedrake/preconditioners/fdm.py]: replaced this function for a triple matrix product

File: firedrake/preconditioners/fdm.py (hunk 36 above)
Comment (pbrubeck) [reply firedrake/preconditioners/fdm.py]: replaced this with variables external to the function

File: .github/workflows/build.yml [hunk 40]
```diff
@@ -48,7 +48,7 @@ jobs:
       - name: Build Firedrake
         run: |
           cd ..
-          ./firedrake/scripts/firedrake-install $COMPLEX --venv-name firedrake_venv --tinyasm --netgen --disable-ssh --minimal-petsc --slepc --documentation-dependencies --install thetis --install gusto --install icepack --install irksome --install femlium --no-package-manager || (cat firedrake-install.log && /bin/false)
+          ./firedrake/scripts/firedrake-install $COMPLEX --venv-name build --tinyasm --disable-ssh --minimal-petsc --slepc --documentation-dependencies --install thetis --install gusto --install icepack --install irksome --install femlium --no-package-manager --package-branch tsfc pbrubeck/fdm-discontinuous || (cat firedrake-install.log && /bin/false)
```
Comment (dham) [inline .github/workflows/build.yml:51]: Don't change `firedrake_venv` to build.

File: firedrake/preconditioners/fdm.py [hunk 41]
```diff
@@ -1,132 +1,168 @@
...
     Here we assume that the volume integrals in the Jacobian can be expressed as:
 
-    inner(grad(v), alpha(grad(u)))*dx + inner(v, beta(u))*dx
+    inner(d(v), alpha * d(u))*dx + inner(v, beta * u)*dx
 
-    where alpha and beta are linear functions (tensor contractions).
-    The sparse matrix is obtained by approximating alpha and beta by cell-wise
-    constants and discarding the coefficients in alpha that couple together
-    mixed derivatives and mixed components.
+    where alpha and beta are possibly tensor-valued.  The sparse matrix is
```
Comment (dham) [inline firedrake/preconditioners/fdm.py:68]: explicitly say what `d` is here.

**Review by dham — COMMENTED** [review]
Yes subject to the tiny points.

**Review by dham — CHANGES_REQUESTED** [review]

**Review by dham — DISMISSED** [review]

**Review by pefarrell — APPROVED** [review]
